### ISSUE TITLE
Big picture shelljs@1.0.0 discussion

### ISSUE THREAD

**Issue by ariporad** [shelljs/shelljs#444]
Hi Everyone,

As discussed in #443, I'd like to start discussing `shelljs@1.0.0` features on a high level.

`v1.0.0` features [@shelljs/contributors, feel free to edit]:

- [ ] Async versions of all commands (#402, #387)
- [ ] Plugin System (#391)
    - [ ] It would be quite nice to impliment all built in commands as plugins.
- [ ] Code Style Guidelines (#317)
- [ ] Miscellaneous cleanup/refactor of everything.

Thoughts?

**Comment by levithomason** [comment]
Yep to all, everything that was on my mind :)

**Comment by nfischer** [comment]
I don't like the idea of moving core commands outside of this package (not sure if that's what you meant by "implement all built-in commands as plugins").

I'm in favor of the other ideas. I'll update with the other things we have on the milestone.

**Comment by ariporad** [comment]
@nfischer: It's not strictly necessary, they could be plugins inside this repo, although I think they should be separate packages for a few reasons:

1. It's more modular, so it's foreseeable that modules could be used independently of shelljs. (Ex, `mkdir -p` could be useful stand-alone.)
2. Imagine that one day `@shelljs/rm@1.5.0` is published with a horrible mistake! It doesn't actually remove anything. But, it shipped with `shelljs@1.5.0`, and many dependencies have already updated their `package.json`! If it's a separate module, we can release `@shelljs/rm@1.5.1`, and when you install `shelljs@1.5.0` it automagically installs `@shelljs/rm@1.5.1`.
3. Then we're 100% dogfooding our API.

**Comment by ariporad** [comment]
@nfischer: ping.

**Comment by levithomason** [comment]
What if this was done similar to lodash.  You get the entire library on install, but you can also simply 
```
npm i lodash.merge
```
if you just want to merge objects.  Everything is kept in one repo, but many packages are published from it's subdirectories.

This would allow people to just install `shelljs.mkdir`, for example, or the entire lib.  Not a priority, but it also opens shelljs up for the same pattern as the [lodash babel plugin](https://github.com/lodash/babel-plugin-lodash).

**Comment by ariporad** [comment]
@levithomason: Agreed. That's along the lines of what I was thinking. Could be really nice.

That would also let someone use (for example), `rimraf` to impliment a custom `$.rm`. (I don't know why they'd need to, but whatever.)

**Comment by levithomason** [comment]
I've made my peace in this one, will let you too wrap it up!

**Comment by nfischer** [comment]
> if you just want to merge objects. Everything is kept in one repo, but many packages are published from it's subdirectories.

Sure, cash does it this way.
